src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,167 @@
+    private currentState: StatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initialState;
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initialState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: StatefulActionButtonState) {
+        if (this.options.states.indexOf(state) === -1) {
+            console.warn(
+                `State '${state.name}' does not exist on this StatefulActionButton\nEnsure to use the object references used at the instantiation.`
+            );
+            return;
+        }
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
+                `Transition from State '${this.currentState.name}' to State '${state.name}' is not allowed on this StatefulActionButton.\nEnsure to use the object references used at the instantiation.`
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initialState) {
+            console.warn('The stateful action button cannot render if the initial states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initialState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        return !this.options.allowedTransitions || this.areTransitionsValid();
+    }
+
+    private areTransitionsValid(): boolean {
+        for (let index = 0; index < this.options.allowedTransitions.length; index++) {
```

Comment:
```
if (
  !this.options.allowedTransitions.every(
    (transition) =>
      this.options.states.includes(transition.from) &&
      this.options.states.includes(transition.to)
  )
) {
  console.warn(
    "All transitions of the StatefulActionButton need to be included as valid states of the button."
  );
  return false;
}
return true;
```

Ou tu veux de la compatibilite plus vieille que ca?

Reply:
J'ai pensé, je suid d'accord qu'il y a de la complexité en plus, mais je voulais donner un log précis a l'utilisateur, car je veux pas qu'il se casse la tête a testé l'un puis l'autre facon USB A. (Is it this oriented properly, turn around, try again, fail, turn around again, succeed (maybe)